File: RemoteInstaller/app/src/main/java/remoteapkinstallerclient/com/br/remoteinstaller/service/AppService.java
package remoteapkinstallerclient.com.br.remoteinstaller.service;

import android.content.Context;

import java.util.List;

import remoteapkinstallerclient.com.br.remoteinstaller.objects.App;
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.APIClient;
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.apps.AppHttpService;
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.apps.AppsResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

/**
 * Created by eduardo.sampaio on 07/12/17.
 */

public class AppService {

    public interface ListAppsServiceCallback{

        void onSuccess(List<App> apps);

        void onError(Throwable cause);
    }

    public void listApplications(Context context, final ListAppsServiceCallback callback){
        Retrofit retrofit = APIClient.getClient(context);
        AppHttpService resquestManager = retrofit.create(AppHttpService.class);

        resquestManager.getApps().enqueue(new Callback<AppsResponse>() {
            @Override
            public void onResponse(Call<AppsResponse> call, Response<AppsResponse> response) {
                if(response.isSuccessful()) {
                    List<App> listApps = response.body().getResponse();
                    callback.onSuccess(listApps);
                }else{
                    callback.onError(new RuntimeException("Error processing request"));
                }
            }

            @Override
            public void onFailure(Call<AppsResponse> call, Throwable throwable) {
                callback.onError(throwable);
            }
        });
    }
}

File: RemoteInstaller/app/src/main/java/remoteapkinstallerclient/com/br/remoteinstaller/service/AppService.kt
package remoteapkinstallerclient.com.br.remoteinstaller.service

import android.content.Context

import remoteapkinstallerclient.com.br.remoteinstaller.objects.App
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.APIClient
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.apps.AppHttpService
import remoteapkinstallerclient.com.br.remoteinstaller.service.http.apps.AppsResponse
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit

/**
 * Created by eduardo.sampaio on 07/12/17.
 */

class AppService {

    public interface ListAppsServiceCallback {

        fun onSuccess(apps: List<App>?)

        fun onError(cause: Throwable)
    }

    fun listApplications(context: Context, callback: ListAppsServiceCallback) {
        val retrofit = APIClient.getClient(context)
        val resquestManager = retrofit?.create(AppHttpService::class.java)

        resquestManager?.apps?.enqueue(object : Callback<AppsResponse> {
            override fun onResponse(call: Call<AppsResponse>, response: Response<AppsResponse>) {
                if (response.isSuccessful) {
                    val listApps = response.body()!!.response
                    callback.onSuccess(listApps)
                } else {
                    callback.onError(RuntimeException("Error processing request"))
                }
            }

            override fun onFailure(call: Call<AppsResponse>, throwable: Throwable) {
                callback.onError(throwable)
            }
        })
    }
}
